omniscidb/ArrowStorage/ArrowStorage.cpp

```diff
@@ -630,6 +630,176 @@ TableInfoPtr ArrowStorage::importArrowTable(std::shared_ptr<arrow::Table> at,
+                                                  const size_t new_frag_size) {
+  if (!new_frag_size) {
+    throw std::runtime_error("Cannot refragment to fragment size 0");
+  }
+
+  if (table_name.empty()) {
+    throw std::runtime_error("Cannot create table with empty name");
+  }
+  TableInfoPtr res;
+  int new_table_id;
+  TableData original_table_data;
+  mapd_unique_lock<mapd_shared_mutex> data_lock(data_mutex_);
+  {
+    mapd_unique_lock<mapd_shared_mutex> dict_lock(dict_mutex_);
+    mapd_unique_lock<mapd_shared_mutex> schema_lock(schema_mutex_);
+    if (getTableInfoNoLock(db_id_, new_table_name)) {
+      throw std::runtime_error("Table with name '"s + new_table_name +
+                               "' already exists"s);
+    }
+
+    auto tinfo = getTableInfoNoLock(db_id_, table_name);
+    if (!tinfo) {
+      throw std::runtime_error("Unknown table: "s + table_name);
+    }
+    auto& table = *tables_.at(tinfo->table_id);
+    original_table_data.schema = table.schema;
+    original_table_data.col_data = table.col_data;
+    original_table_data.row_count = table.row_count;
+    original_table_data.table_stats = table.table_stats;
+    new_table_id = next_table_id_++;
+    res = addTableInfo(db_id_, new_table_id, new_table_name, false, 0, 0);
+    auto col_info_list = listColumnsNoLock(db_id_, tinfo->table_id);
+    for (size_t col_idx = 0; col_idx < col_info_list.size() - 1; col_idx++) {
+      const auto col_info = col_info_list[col_idx];
+      CHECK(col_info);
+      auto col_type = col_info->type;
+      auto col_arr = table.col_data[col_idx];
+      addColumnInfo(
+          db_id_, new_table_id, columnId(col_idx), col_info->name, col_type, false);
+      if (col_type->isExtDictionary() && col_arr->type()->id() == arrow::Type::STRING) {
+        auto dict_data =
+            dicts_.at(dynamic_cast<const hdk::ir::ExtDictionaryType*>(col_type)->dictId())
+                .get();
+        CHECK(dict_data);
+        dict_data->addTableColumnPair(new_table_id, col_idx);
+      }
+    }
+  }
+  auto [iter, inserted] = tables_.emplace(new_table_id, std::make_unique<TableData>());
+  CHECK(inserted);
+
+  auto& new_table = *iter->second;
+  mapd_unique_lock<mapd_shared_mutex> table_lock(new_table.mutex);
+  data_lock.unlock();
+  new_table.schema = original_table_data.schema;
+  new_table.col_data = original_table_data.col_data;
+  new_table.row_count = original_table_data.row_count;
+  new_table.table_stats = original_table_data.table_stats;
+
+  refragmentTable(new_table, new_table_id, new_frag_size);
+  return res;
+}
+
+void ArrowStorage::refragmentTable(TableData& table,
+                                   const int table_id,
+                                   const size_t new_frag_size) {
+  if (!new_frag_size) {
+    throw std::runtime_error("Cannot refragment to fragment size 0");
+  }
+  const size_t new_frag_count =
+      std::ceil(static_cast<float>(table.row_count) / new_frag_size);
```

Comment:
Float precision is not good enough here. If you have 1'000'000'001 rows and fragment size is 10'000'000 then your result here would be 100 fragments, not 101. Use integer arithmetics to avoid precision issues.